### DISCUSSION TITLE
关于数据可靠性的疑问

### DISCUSSION THREAD

**Discussion by XLPE** [alibaba/ilogtail#394] [Help] [answered]
您好，看了社区文章被ilogtail的很多优秀的特性所吸引。应内部需求，也扩展了flusher部分，对接pulsar。
其中对应社区文章，我总结ilogtail有下面一些优势：
![image](https://user-images.githubusercontent.com/16996754/190362598-3731efed-a7cb-41d7-9d7e-cf76ef958a70.png)
不过深入研究和测试过程中，发现仅仅扩展flusher，无法做到上图中的可靠性。发送状态没有和采集的checkpoint关联起来。
而在进程可靠性上，由于缺少守护进程，也无法做到异常自动恢复。

对于flusher部分，sls实际是基于C++实现的。对于自己扩展的flusher，官方是否也建议扩展C++呢？
不过过了下代码，基本上特性都已经开源了，只是一些特性集成在sls代码里面。如果是对接新的flusher，可能需要定制一下才行。

**Comment by messixukejia** [comment, marked as answer]
很感谢能够在ilogtail社区提出宝贵的建议。关于上述问题：
1、守护进程这个目前没有开源，企业版是有这部分能力的，基本原理就是父进程fork出ilogtail进程，ilogtail退出后负责拉起。可以按照这个原则实现下，贡献给ilogtail社区。
2、flusher部分涉及上下游的联动，特别是流量控制、采集checkpoint关联等对于每个flusher来分别实现确实会比较复杂，后面我们提供公共的接口，各flusher可以直接通过接口来对接。
3、扩展的flusher优先还是以go为主，除非有特定的需求，后续再考虑支持c++ flusher。
